be/src/olap/rowset/segment_v2/segment_iterator.cpp

```diff
@@ -1632,19 +1685,35 @@ void SegmentIterator::_init_current_block(
     for (size_t i = 0; i < _schema->num_column_ids(); i++) {
         auto cid = _schema->column_id(i);
         auto column_desc = _schema->column(cid);
-        // the column in block must clear() here to insert new data
-        if (_is_pred_column[cid] ||
-            i >= block->columns()) { //todo(wb) maybe we can release it after output block
-            current_columns[cid]->clear();
-        } else { // non-predicate column
-            current_columns[cid] = std::move(*block->get_by_position(i).column).mutate();
-
-            if (column_desc->type() == FieldType::OLAP_FIELD_TYPE_DATE) {
-                current_columns[cid]->set_date_type();
-            } else if (column_desc->type() == FieldType::OLAP_FIELD_TYPE_DATETIME) {
-                current_columns[cid]->set_datetime_type();
-            }
+        if (!_is_pred_column[cid] &&
+            !_segment->same_with_storage_type(
```

Comment:
Now it's safe, since only variant's predicate will continue here, and let the expr to handle such filter like inverted index match does.It could be a common usage for later light schema change data type

Reply:
Add CHECK in `same_with_storage_type`  to make sure continue in variant case